crates/providers/src/heart.rs

```diff
@@ -0,0 +1,283 @@
+    /// the latest block.
+    fn handle_new_block(&mut self, block: Block, latest: &watch::Sender<Option<Block>>) {
+        // Blocks without numbers are ignored, as they're not part of the chain.
+        let Some(block_height) = &block.header.number else { return };
+
+        // Check if we are watching for any of the transactions in this block.
+        let to_check =
+            block.transactions.hashes().filter_map(|tx_hash| self.unconfirmed.remove(tx_hash));
+        for watcher in to_check {
+            // If `confirmations` is 0 we can notify the watcher immediately.
+            let confirmations = watcher.config.confirmations;
+            if confirmations == 0 {
+                watcher.notify();
+                continue;
+            }
+            // Otherwise add it to the waiting list.
+            debug!(tx=%watcher.config.tx_hash, %block_height, confirmations, "adding to waiting list");
+            self.waiting_confs
+                .entry(*block_height + U256::from(confirmations))
+                .or_default()
+                .push(watcher);
+        }
+
+        self.check_confirmations(block_height);
+
+        // Update the latest block. We use `send_replace` here to ensure the
+        // latest block is always up to date, even if no receivers exist.
+        // C.f. https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send
+        debug!(%block_height, "updating latest block");
+        let _ = latest.send_replace(Some(block));
+    }
+}
+
+impl<S: Stream<Item = Block> + Unpin + Send + 'static> Heartbeat<S> {
+    /// Spawn the heartbeat task, returning a [`HeartbeatHandle`]
+    pub(crate) fn spawn(mut self) -> HeartbeatHandle {
+        let (latest, latest_rx) = watch::channel(None::<Block>);
+        let (ix_tx, mut ixns) = mpsc::channel(16);
+
+        let fut = async move {
+            'shutdown: loop {
+                {
+                    // We bias the select so that we always handle new messages
+                    // before checking blocks, and reap timeouts are last.
+                    let next_reap = self.next_reap();
+                    select! {
+                        biased;
+
+                        // Watch for new transactions.
+                        ix_opt = ixns.recv() => match ix_opt {
+                            Some(to_watch) => self.handle_watch_ix(to_watch),
+                            None => break 'shutdown, // ix channel is closed
+                        },
+
+                        // Wake up to handle new blocks.
+                        block = self.stream.select_next_some() => {
+                            self.handle_new_block(block, &latest);
+                        },
+
+                        // This arm ensures we always wake up to reap timeouts,
+                        // even if there are no other events.
+                        _ = tokio::time::sleep_until(next_reap.into()) => {},
```

Comment:
hmm, I don't think this is correct because Sleep is not Unpin.
 
I believe this should be pinned:

https://docs.rs/tokio/latest/tokio/time/struct.Sleep.html#examples